src/server/dfly_main.cc

```diff
@@ -456,103 +456,135 @@ bool ShouldUseEpollAPI(const base::sys::KernelVersion& kver) {
-
-  // strip 0::<path> into just <path>, and newline.
   string cgv = std::move(cg.value());
-  size_t pos = cgv.rfind(':');
-  if (pos == string::npos)
-    return nonstd::make_unexpected(make_error_code(errc::not_supported));
 
-  string res = cgv.substr(pos + 1);
-  absl::StripTrailingAsciiWhitespace(&res);
+  // Next, depending on cgroup version we either read:
+  // N:<cgroup name>:<path> -- in case of v1, in many lines
+  // 0::<cgroup name> -- in case of v2, in a single line
 
-  return res;
+  auto stripped = absl::StripAsciiWhitespace(cgv);
+
+  vector<string_view> groups = absl::StrSplit(stripped, '\n');
+
+  if (groups.size() == 1) {
+    // for v2 we only read 0::<name>
+    size_t pos = cgv.rfind(':');
+    if (pos == string::npos)
+      return make_error_code(errc::not_supported);
+
+    string res = cgv.substr(pos + 1);
+    absl::StripTrailingAsciiWhitespace(&res);
+
+    *memory_path = "/sys/fs/cgroup/" + res;
+    *thread_path = "/sys/fs/cgroup/" + res;
+  } else {
+    for (const auto& sv : groups) {
+      // in v1 the format is
+      // N:s1:2 where N is an integer, s1, s2 strings with s1 maybe empty.
+      vector<string_view> entry = absl::StrSplit(sv, ':');
+      CHECK(entry.size() == 3);
+
+      if (entry[1] == "memory")
+        *memory_path = absl::StrCat("/sys/fs/cgroup/memory/", entry[2]);
+
+      if (entry[1] == "cpu,cpuacct")
+        *memory_path = absl::StrCat("/sys/fs/cgroup/cpu,cpuacct", entry[2]);
+    }
+  }
+
+  return error_code{};
 }
 
 void UpdateResourceLimitsIfInsideContainer(io::MemInfoData* mdata, size_t* max_threads) {
   /**
-   * (attemps) to check whether we are running
-   * inside a container or not.
-   *
-   * We employ several tests, all of which
-   * are flawed, however together do cover a
-   * good portion of cases.
-   *
-   * 1. Checking '/.dockerenv': very simple, however
-   * only works for Docker, and this file may be moved
-   * in the future.
-   * 2. Checking '/sys/fs/cgroup/memory.max': This
-   * directory -cannot- be edited (not even using sudoedit),
-   * so users are not able to add files to there. The file
-   * 'memory.max' should contain a memory limit for processes.
-   *
-   * We also use this file to find out how much memory we can use.
-   * However, on LXC this is placed in a different directory:
-   * 3. Checking '/sys/fs/cgroup/memory/memory.limit_in_bytes':
-   * Same idea.
+   * Initially check whether we are under v1 or v2.
+   * In v2, /self/proc/cgroup would only have one line, 0::<cgroup name>
+   * In v1, this file would usually have many lines.
    */
 
   using io::Exists;
 
-  auto cgroup_res = GetCGroupPath();
-  if (!cgroup_res) {
-    VLOG(1) << "Failed to get cgroup path, error: " << cgroup_res.error().message();
-    return;
-  }
+  //  auto cgroup_res = GetCGroupPath();
+  //  if (!cgroup_res) {
+  //    VLOG(1) << "Failed to get cgroup path, error: " << cgroup_res.error().message();
+  //    return;
+  //  }
```

Comment:
Please remove instead of commenting out

Reply:
Fixed in the newer version! :slightly_smiling_face: